packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,326 @@
+    }
+}
+```
+
+Here we have a column visualized in different ways based on custom logic:
+
+```TS
+type TableColumnPositiveNegativeNumberCellData = NumberColumnData<'value'>;
+
+const isPositive = (value: number): bool => {
+    return value >= 0;
+}
+
+public TableColumnPositiveNegativeNumber extends FoundationElement implements ITableColumn<TableColumnPositiveNegativeNumberCellData> {
+    ...
+
+    public cellDataKeyNames = ['value'] as const;
+
+    @attr
+    public valueKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public readonly cellStyles: ElementStyles = 
+        css`
+            .good {
+                color: green;
+            }
+
+            .bad {
+                color: red;
+            }
+        `;
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnPositiveNegativeNumberCellData>> =
+        html<TableCellState<TableColumnPositiveNegativeNumberCellData>`
+            html`
+                <nimble-text-field
+                    class=${_ => isPositive(x.data.value) ? "good" : "bad"}
+                    readonly="true"
+                    value=${x => x.data.value}
+                    style="color: red;"
+                >
+                </nimble-text-field>
+            `}
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['value'] === 'number';
+    }
+}
+```
+
+Finally, here is a column that allows a user to register a callback for a click event on a button inside the cell template:
+
+```TS
+type TableColumnButtonCellData = StringColumnData<'id'>;
+
+const fireEvent = (buttonElement: HTMLElement, clickData: { data: string}): void => {
+    const event = new CustomEvent('button-click', { detail: data });
+    buttonElement.dispatchEvent(event);
+}
+
+public TableColumnButton extends FoundationElement implements ITableColumn<TableColumnButtonCellData> {
+    ...
+
+    public cellDataKeyNames = ['id'] as const;
+
+    @attr
+    public idKey: string;
+
+    public getRowDataKeys(): string[] {
+        return [valueKey];
+    }
+
+    public callback: (id: string) => void;
+
+    public readonly cellTemplate: ViewTemplate<TableCellState<TableColumnButtonCellData>> =
+        html<TableCellState<TableColumnButtonCellData>>`
+            <nimble-button @click="${(x, c) => fireEvent(c.event.currentTarget, {data: x.data.id})}>
+                <span>Press Me</span>
+            </nimble-button>
+        `;
+
+    public validateCellData(cellData: TCellData): boolean {
+        return typeof cellData['id'] === 'string';
+    }
+}
+```
+
+Angular template:
+
+```HTML
+<nimble-table #table>
+    <nimble-table-column-button></nimble-table-column-button>
+</nimble-table>
+```
+
+Angular code:
+
+```TS
+@ViewChild('table', { read: NimbleTable, static: true }) private readonly table!: NimbleTable;
+
+public ngOnInit(): void {
+    table.addEventListener('button-click', doSomething)
```

Comment:
I think you should be able to do normal angular event binding in the template itself. At least that should be our expected goal.

It may be that each plugin that is firing events will have to include a directive to listen for the events and turn them to `@Input`s and get good typing of the event details.

Reply:
I agree with @rajsite's suggestion. I'll nit-pick it a bit to say that I would expect the event to be an `@Output`, not an `@Input`.

Reply:
Updated the example code to make it use an expecting event binding syntax (no `addEventListener`), and added a note on how the ultimate APIs that would be responsible for getting that working are TBD.